src/librustc_resolve/late.rs

```diff
@@ -351,6 +351,9 @@ struct LateResolutionVisitor<'a, 'b> {
 
     /// Only used for better errors on `fn(): fn()`.
     current_type_ascription: Vec<Span>,
+
+    /// Only used for better errors on `let <pat>: <expr, not type>;`.
+    current_let_binding: Option<(Span, Span)>,
```

Comment:
At this point there are a lot of "for diagnostics" state in `LateResolutionVisitor` -- could we perhaps move these into their own struct `diag: Diagnostics` to clarify this throughout the code so that they don't seep into semantic logic?